gcsfs/core.py

```diff
@@ -335,15 +335,18 @@ def close_session(loop, session):
         if loop is not None and session is not None:
             if loop.is_running():
                 try:
-                    loop = asyncio.get_event_loop()
```

Comment:
This is essentially "are we in the same thread as the loop".
I don't believe create_task should be done from other threads (this is what loop.call_soon_threadsafe or asyncio.run_coroutine_threadsafe are for, used by sync() just below)

Reply:
A bit more detail on the symptom that I'm seeing

* We hit a `RuntimeError` in both `create_task()` and `sync()` because the `loop` (as returned by `asyncio.get_event_loop()` is closed.
* The `loop` parameter to the function is not the same as `asyncio.get_event_loop()` (calling `id()` on them yields different results).

Sorry, I'm not quite following the intent here. Is the goal to run `session.close` on the event loop of whatever the current thread is? If so, maybe we should guard things by checking running and closed status on the loop returned by `asyncio.get_event_loop()` instead of the function parameter.

Reply:
The gcsfs instance should be used in async mode os sync mode, not both. If the former, we will always be within the event loop, it will be running and we will be on its thread. If sync, the loop should never be in the current thread.
Do you actually have multiple loops in your application?

Reply:
The exception is garbage collection, which happens on the main thread, and the original loop/thread may or may not be live

Reply:
... and garbage collection maybe well happen during interpreter shutdown, when any given object attribute might have disappeared. I am not certain why the session really need to be shut down in such a case, but not doing so produces warnings.

Reply:
Yes, this is happening in the weakref callback which is presumably invoked from GC.

This is what I see with fsspec 2023.9.2 and gcsfs 2023.9.2 (i.e. without my PR applied):

```
site-packages/gcsfs/core.py:342: RuntimeWarning: coroutine 'ClientSession.close' was never awaited

  pass

RuntimeWarning: Enable tracemalloc to get the object allocation traceback

Traceback (most recent call last):

  File "lib/python3.10/weakref.py", line 667, in _exitfunc

    f()

  File "lib/python3.10/weakref.py", line 591, in __call__

    return info.func(*info.args, **(info.kwargs or {}))

  File "site-packages/gcsfs/core.py", line 344, in close_session

    sync(loop, session.close, timeout=0.1)

  File "site-packages/fsspec/asyn.py", line 76, in sync

    raise RuntimeError("Loop is not running")

RuntimeError: Loop is not running
```

Reply:
I am using the synchronous interface to fsspec

```
    def my_function():
        fs = fsspec.filesystem("gcs")
        fs.invalidate_cache()

        if not fs.exists(f"gs://..."):
            return

        for path in fs.glob(
            f"gs://.../*/*/*"
        ):
            # Do some stuff
```

Reply:
Ok, how about this

Leave the `create_task` section running against the current thread's event loop (which can `RuntimeError` if there is none).

Change the `sync` section to run against the `loop` parameter as passed to this function.

Reply:
That sounds reasonable. It may also be worth digging how the two values end up different.

Reply:
My guess --

* The `loop` parameter is the loop owned by the fsspec IO thread (our `GCSFileSystem` instance is created w/o an explicit `loop`, so we get the default).
* Because we're running in the weakref GC callback, we're in the main thread.